TileMap.gd

```diff
@@ -66,265 +82,614 @@ func _on_Player_collided(collision):
 				overworld.score += 1000
 				overworld.win_event()
 	
-func is_cell_vacant(pos, direction):
-	var x = (pos.x / tile_size.x) + direction.x
-	var y = (pos.y / tile_size.y) + direction.y
-	
-	var tile_type = Tile.Stone	
-	if x > 0 && x < level_size.x && y >= 0 && y < level_size.y:
-		tile_type = map[x][y]
-	print("The tile is: " + str(Tile.keys()[tile_type]))
-	match tile_type:
-		Tile.Floor:
-			return true
-		Tile.Door:
-			set_tile(x, y, Tile.Floor)
-			return false
-		Tile.Wall:
-			return false
-		Tile.Stone:
-			return false
-		Tile.Ladder:
-			level_num += 1
-			overworld.score += 20
-			if level_num < LEVEL_SIZES.size():
-				build_level()
-				overworld.place_player()
-			else:
-				overworld.score += 1000
-				overworld.win_event()
-	
-func update_child_pos(node):
-	var current_pos = node.position / tile_size
-	var new_pos = current_pos + node.direction
-	var target_pos = new_pos * tile_size
-	return target_pos
-
 func build_level():
-	# start with blank map
-	rooms.clear()
-	map.clear()
 	self.clear()
 	
 	level_size = LEVEL_SIZES[level_num]
-	for x in range(level_size.x):
-		map.append([])
-		for y in range(level_size.y):
-			map[x].append(Tile.Stone)
-			self.set_cell(x, y, Tile.Stone)
-			
-	var free_regions = [Rect2(Vector2(2,2), level_size - Vector2(4,4))]
-	var num_rooms = LEVEL_ROOM_COUNTS[level_num]
-	for i in range(num_rooms):
-		add_room(free_regions)
-		if free_regions.empty():
-			break
-	
-	connect_rooms()
+	#populate possible starting points
+	generate_possible_start_points()
+	#populate all top left corner vectors of the 9x9 cells	
+	generate_Anchor_Points()
+	#populate door/opening midpoints
+	generate_midpoints_for_doors_openings()
+	#populate map corners array
+	generate_Map_Corner_Points()
+	
+	#Max number of rooms based on RoomCellSizes and level size
+	var numberOfCells = (level_size.x / RoomCellSize.x) * (level_size.y / RoomCellSize.y)
+	var startingSpot = possibleStartPoints[randi() % possibleStartPoints.size()]
+	var firstSpot = true
+	var needsCorridor = false
+		
+	for i in range(numberOfCells):
+#		yield(get_tree().create_timer(0.5), 'timeout')
+		var currentSpot
+		if firstSpot:
+			currentSpot = startingSpot
+			firstSpot = false
+		else:
+			currentSpot = nextTileCellQueue.pop_front()
+		
+		#add the room
+		add_room(currentSpot, needsCorridor)
 
-	# Place end ladder
-	var end_room = rooms.back()
-	var ladder_x = end_room.position.x + 1 + randi() % int(end_room.size.x - 2)
-	var ladder_y = end_room.position.y + 1 + randi() % int(end_room.size.y - 2)
-	set_tile(ladder_x, ladder_y, Tile.Ladder)
+		if needsCorridor:
+			#close blocked off sections
+			#TODO: number of exits is based on corridor scene will need to get that value somewhere
```

Comment:
If this TODO is an issue, you should create an issue for it. Otherwise we should delete this comment.